File: hw/hw06/hw06.py
#!/usr/bin/env python
# coding: utf-8

# In[238]:


from typing import List

def threeSum( nums: List[int]) -> List[List[int]]:
    fullList = []
    setList = []
    nums.sort()
    print(nums)
    for x in range(len(nums)-2):
       #print(x)
        for y in range(x+1,len(nums)):
            #print(y)
            for z in range(y+1,len(nums)):
                #print(z)
                #print(f"{nums[x]}:{nums[y]}:{nums[z]}")
                combo = {nums[x],nums[y],nums[z]}
                if nums[x] + nums[y] + nums[z] == 0 and combo not in setList:
                    setList.append(combo)
                    fullList.append([nums[x],nums[y],nums[z]])
    return fullList
        


# In[239]:


from typing import List

def threeSum( nums: List[int]) -> List[List[int]]:
    fullList = []
    nums.sort()
    
    for i in range(len(nums)-2):
       if i > 0 and nums[i] == nums[i - 1]:
            continue
       left,right = i + 1, len(nums)-1 
       while left < right:
            total = nums[i] + nums[left] + nums[right]
            if total < 0:
               left+=1
            elif total >0:
               right-=1
            else:
                fullList.append([nums[i],nums[left],nums[right]])
                left +=1
                right -=1
                while left < right and nums[left] == nums[left - 1]:
                    left += 1
                while left < right and nums[right] == nums[right + 1]:
                    right -= 1

    return fullList
        


# In[240]:


nums = [0,1,1]
threeSum(nums)


# In[241]:


nums = [-5,0,5,10,-10,0] 
threeSum(nums)


# 
# def threeSum(self, nums: List[int]) -> List[List[int]]:
# 
#   #Your code here
# 
# Announced Test Cases:
# 
# Input: nums = [0,1,1]
# Output: []
# Explanation: The only possible triplet does not sum up to 0.
# 
# Input: nums = [-5,0,5,10,-10,0] 
# Output: [[-10,0,10],[-5,0,5]] 
# Explanation: There are two possible combinations of triplets that satisfy: (-5,0,5) and (-10,0,10).
# 
# Hint: There are 3 well-known ways to solve this problem!
# 
# Hint 2: This problem should remind you of another homework problem that you hopefully have already solved. 
# 
# Hint 3: This problem might benefit from the algorithms we have recently discussed in class!

# In[ ]:


get_ipython().system('jupyter nbconvert --to script hw06.ipynb')

File: hw06.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class hw06 {

    // Approach 1: Using nested loops
    public static List<List<Integer>> threeSum1(int[] nums) {
        List<List<Integer>> fullList = new ArrayList<>();
        Set<List<Integer>> setList = new HashSet<>();
        Arrays.sort(nums);

        for (int x = 0; x < nums.length - 2; x++) {
            for (int y = x + 1; y < nums.length; y++) {
                for (int z = y + 1; z < nums.length; z++) {
                    List<Integer> combo = new ArrayList<>();
                    combo.add(nums[x]);
                    combo.add(nums[y]);
                    combo.add(nums[z]);
                    if (nums[x] + nums[y] + nums[z] == 0 && !setList.contains(combo)) {
                        setList.add(combo);
                        fullList.add(combo);
                    }
                }
            }
        }
        return fullList;
    }

    // Approach 2: Using two-pointer technique
    public static List<List<Integer>> threeSum2(int[] nums) {
        List<List<Integer>> fullList = new ArrayList<>();
        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int left = i + 1, right = nums.length - 1;
            while (left < right) {
                int total = nums[i] + nums[left] + nums[right];
                if (total < 0) {
                    left++;
                } else if (total > 0) {
                    right--;
                } else {
                    fullList.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    left++;
                    right--;
                    while (left < right && nums[left] == nums[left - 1]) {
                        left++;
                    }
                    while (left < right && nums[right] == nums[right + 1]) {
                        right--;
                    }
                }
            }
        }
        return fullList;
    }

    public static void main(String[] args) {
        int[] nums = {0, 1, 1};
        System.out.println("Approach 1:");
        System.out.println(threeSum1(nums));  // Output: []

        System.out.println("Approach 2:");
        System.out.println(threeSum2(nums));  // Output: []

        int[] nums2 = {-5, 0, 5, 10, -10, 0};
        System.out.println("Approach 1:");
        System.out.println(threeSum1(nums2));  // Output: [[-10, 0, 10], [-5, 0, 5]]

        System.out.println("Approach 2:");
        System.out.println(threeSum2(nums2));  // Output: [[-10, 0, 10], [-5, 0, 5]]
    }
}
